File: java/src/main/java/QuestionEight.java
import java.util.HashSet;

public class QuestionEight {

	// Runtime Complexity O(N) Space Complexity O(N)
	public static int indexOfFirstDuplicate(int[] numbers) {
		HashSet<Integer> set = new HashSet<>();
		for (int i = 0; i < numbers.length; i++) {
			int number = numbers[i];
			if (set.contains(number)) {
				return i;
			}
			set.add(number);
		}
		return -1;
	}
}

File: go/questions/question_eight.go
package questions

// IndexOfFirstDuplicate Runtime Complexity O(N) Space Complexity O(N)
func IndexOfFirstDuplicate(numbers []int) int {
	numberMap := make(map[int]bool)
	for index, number := range numbers {
		if _, ok := numberMap[number]; ok {
			return index
		} else {
			numberMap[number] = true
		}
	}
	return -1
}
